### PR TITLE
feat(ffi): expose filtered replication via p2p_add_replicator_with_filter and defra_mobile_add_replicator

### LINKED ISSUE #1075: FFI: expose filtered replication (p2p_add_replicator_with_filter / mobile replicator wrapper)
## Summary

Filtered P2P replication is not reachable through the FFI surface. The underlying P2P adapter (`add_replicator`) already supports per-collection filters, and the HTTP API already accepts them, but the C FFI does not expose them. There is also no mobile convenience wrapper for adding a replicator at all.

This blocks local-first mobile clients (e.g. an iOS app embedding DefraDB via the v0.15 `defra-ffi-iroh` XCFramework) from setting up scoped/conversation-filtered replication against a remote DefraDB node.

## Current behavior

`crates/ffi/src/p2p/replicator.rs` — `p2p_add_replicator` passes an empty filter map:

```rust
p2p.system.ops().add_replicator(collections, Some(&addr), Default::default(), Vec::new(), None).await
```

The `filters` argument is hardcoded to `Default::default()` (empty), so every FFI replicator is unfiltered.

`crates/ffi/src/mobile.rs` exposes `defra_mobile_sync_collection`, `defra_mobile_connect`, etc., but has **no** `defra_mobile_add_replicator` convenience wrapper — so the mobile Swift surface cannot add a replicator without dropping to the raw `p2p_add_replicator` symbol, and even then it cannot pass filters.

## Expected behavior

Be able to add a replicator with per-collection filters from the FFI/mobile surface, mirroring the HTTP request shape (`ReplicatorRequest.Filters` → `ReplicationFilters = BTreeMap<String, ReplicationFilter>`):

```json
{"AgentRequest": {"predicate": {"agent_did": {"_eq": "did:key:z..."}}}}
```

## Suggested fix

- Add a new non-breaking FFI symbol, e.g. `p2p_add_replicator_with_filter(node_ptr, identity_did, peer_addr, collections_json, filters_json)`, that parses `filters_json` into `ReplicationFilters` (keyed by collection name) and forwards them to `add_replicator`.
- Optionally add a `defra_mobile_add_replicator(node_ptr, request_json)` convenience wrapper that bundles collections + address + filters (consistent with the existing `defra_mobile_sync_collection` request-JSON style), so the mobile binding gets a single entry point.
- Keep `p2p_add_replicator` as-is (empty filters) for ABI stability.

## Why it matters

For a mobile supervisor/agent client, unfiltered replication would sync the entire remote node's collections. The app needs to scope replication to a specific agent/session/conversation subset (e.g. `agent_did`), which requires the filter argument that already exists in the adapter but is not plumbed through the FFI.

## Environment

- defradb.rs (current `main`)
- Reproduced while building against the v0.15.0 `defra-ffi-iroh_ios_xcframework` artifact.

### PR DESCRIPTION
**Pull request by noahsaso**
Closes #1075.

## Problem
Filtered P2P replication is available through the HTTP API and the underlying P2P adapter, but not through the C FFI. `p2p_add_replicator` keeps using an empty filter map, and mobile embedding did not have a convenience wrapper for adding a replicator.

## Approach
- Added the non-breaking C symbol `p2p_add_replicator_with_filter(node_ptr, identity_did, peer_addr, collections_json, filters_json)` while leaving `p2p_add_replicator` unchanged for ABI stability.
- Added `parse_filters_json` for HTTP-compatible filter JSON, with `null`, `{}`, and `""` treated as unfiltered.
- Added `defra_mobile_add_replicator(node_ptr, request_json)` with camelCase request JSON: `collections`, `peerAddr`, optional `identityDid`, and optional `filters`.
- Re-exported the new FFI and mobile symbols.

## Filters wire format
`filters_json` mirrors the HTTP `Filters` field: a map keyed by collection name. Each value is the HTTP replication filter shape:

```json
{
  "Users": {"Conditions": {"agent_did": {"_eq": "did:key:z..."}}}
}
```

The legacy scalar equality shape is also accepted:

```json
{
  "Users": {"Field": "agent_did", "Value": "did:key:z..."}
}
```

## Validation
- `cargo fmt -p ffi -- --check`: ✅
- `cargo test -p ffi --lib parse_filters`: ✅
- `cargo test -p ffi --lib mobile_add_replicator_request`: ✅
- `cargo build -p ffi --lib`: ✅
- `cargo clippy -p ffi --lib`: ✅

Not covered here: a live two-node P2P dial/replication integration test or a full mobile XCFramework rebuild.

### CODE REVIEW

**Comment by jackzampolin** [comment]
Thanks for plumbing this through. The direction looks right, but I think there are a couple of mobile/API edge cases worth tightening before merge:

- `MobileAddReplicatorRequest.filters` is currently a non-optional `ReplicationFilters` with `#[serde(default)]`. Missing `filters` works, but explicit `"filters": null` appears to fail deserialization before it reaches the lower parser, even though the raw FFI parser treats `null` as unfiltered. Can we accept explicit null here too, either with `Option<ReplicationFilters>` or a null-to-default deserializer, plus a regression test?
- The issue body shows a lowercase `predicate` example, and P2P peerstore JSON also serializes predicate filters that way, while this PR accepts HTTP `Conditions` and legacy `Field`/`Value`. Can we either support `predicate` as an alias or update the docs/issue-facing examples so mobile callers do not copy a shape this endpoint rejects?

Also, since this is meant to unblock mobile embedding: do we need any other mobile-facing bits in this PR before merge, like updating `tools/apple/README.md`, adding a Swift/request JSON example, or documenting that the next XCFramework rebuild will pick up the new symbols?

**Comment by noahsaso** [comment]
Thanks for the review @jackzampolin — all three points are well taken. I'll address them:

1. **Explicit `null` filters**: You're right — `#[serde(default)]` only covers a *missing* field, so `{"filters": null}` fails before reaching the lower-level parser. I'll switch `MobileAddReplicatorRequest.filters` to `Option<ReplicationFilters>` (or a null-to-default deserializer) and add a regression test.

2. **`predicate` shape**: Good catch. The p2p `ReplicationFilter` serializes as `{"predicate": {...}}` (snake_case), but the HTTP type this PR uses only accepts `{"Conditions": {...}}` or legacy `{"Field": "...", "Value": ...}`. I'll add `predicate` as an accepted alias so mobile callers can use either shape, and clarify the accepted formats in the docs.

3. **Mobile docs**: I'll add a Swift/request JSON example and note the XCFramework rebuild in the relevant docs.

Will push an update shortly.

File: crates/ffi/src/mobile_config.rs [hunk 1]
```diff
@@ -81,6 +81,17 @@ pub(crate) struct MobileSyncRequest {
...
 
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct MobileAddReplicatorRequest {
+    pub identity_did: Option<String>,
+    pub collections: Vec<String>,
+    #[serde(alias = "address")]
+    pub peer_addr: String,
+    #[serde(default)]
+    pub filters: defra_http::router::ReplicationFilters,
```
[the remark this answers is not in the document]
Comment (noahsaso) [reply crates/ffi/src/mobile_config.rs]: Fixed in 01582fe by making `MobileAddReplicatorRequest.filters` optional and adding a regression test for explicit `"filters": null`. The mobile wrapper now serializes `None` as `null` when delegating to the lower-level FFI parser.

**Comment by noahsaso** [comment]
Addressed the review feedback in 01582fe:

- `MobileAddReplicatorRequest.filters` now accepts omitted or explicit `null` filters, with a regression test for `"filters": null`.
- The HTTP replication filter type now accepts lowercase `predicate` as an alias for `Conditions`, with a `parse_filters_json` regression test.
- `tools/apple/README.md` now lists `defra_mobile_add_replicator`, includes request JSON examples for `predicate`, `Conditions`, and legacy `Field`/`Value`, and notes that the XCFramework should be rebuilt to pick up the new symbol.

Validation:
- `cargo fmt -p ffi -p defra-http -- --check`
- `cargo test -p ffi --lib parse_filters`
- `cargo test -p ffi --lib mobile_add_replicator_request`
- `cargo test -p ffi --lib`
- `cargo build -p ffi --lib`
- `cargo clippy -p ffi --lib`
